src/import/pull-tar.c

```diff
@@ -346,59 +395,92 @@ static void tar_pull_job_on_finished(PullJob *j) {
+        } else {
+                r = tar_pull_determine_path(i, NULL, &i->final_path);
+                if (r < 0)
+                        goto finish;
 
-                i->temp_path = mfree(i->temp_path);
-
-                if (i->settings_job &&
-                    i->settings_job->error == 0) {
-
-                        /* Also move the settings file into place, if it exists. Note that we do so only if we also
-                         * moved the tar file in place, to keep things strictly in sync. */
-                        assert(i->settings_temp_path);
+                if (!i->tar_job->etag_exists) {
+                        /* This is a new download, verify it, and move it into place */
 
-                        /* Regenerate final name for this auxiliary file, we might know the etag of the file now, and
-                         * we should incorporate it in the file name if we can */
-                        i->settings_path = mfree(i->settings_path);
+                        assert(i->temp_path);
+                        assert(i->final_path);
 
-                        r = tar_pull_determine_path(i, ".nspawn", &i->settings_path);
-                        if (r < 0)
-                                goto finish;
+                        tar_pull_report_progress(i, TAR_FINALIZING);
 
-                        r = import_make_read_only(i->settings_temp_path);
+                        r = import_mangle_os_tree(i->temp_path);
                         if (r < 0)
                                 goto finish;
 
-                        r = rename_noreplace(AT_FDCWD, i->settings_temp_path, AT_FDCWD, i->settings_path);
+                        r = install_file(
+                                        AT_FDCWD, i->temp_path,
+                                        AT_FDCWD, i->final_path,
+                                        INSTALL_READ_ONLY|
+                                        (i->flags & PULL_SYNC ? INSTALL_SYNCFS : 0));
                         if (r < 0) {
-                                log_error_errno(r, "Failed to rename settings file to %s: %m", i->settings_path);
+                                log_error_errno(r, "Failed to rename to final image name to %s: %m", i->final_path);
                                 goto finish;
                         }
 
-                        i->settings_temp_path = mfree(i->settings_temp_path);
+                        i->temp_path = mfree(i->temp_path);
+
+                        if (i->settings_job &&
+                            i->settings_job->error == 0) {
+
+                                /* Also move the settings file into place, if it exists. Note that we do so only if we also
+                                 * moved the tar file in place, to keep things strictly in sync. */
+                                assert(i->settings_temp_path);
+
+                                /* Regenerate final name for this auxiliary file, we might know the etag of the file now, and
```

Comment:
etag?

Reply:
HTTP etag:

https://en.wikipedia.org/wiki/HTTP_ETag

this is an http client, so context should be clear enough